File: scripts/batch-csv-upload.js
/**
 * Batch CSV Upload Script for ProjectLoopbreaker
 * 
 * This script allows you to upload large CSV files in smaller batches
 * to avoid timeouts and make the process more manageable.
 * 
 * Usage:
 *   node scripts/batch-csv-upload.js <path-to-csv> [options]
 * 
 * Options:
 *   --batch-size <number>   Number of rows per batch (default: 10)
 *   --delay <ms>            Delay between batches in milliseconds (default: 1000)
 *   --api-url <url>         API base URL (default: http://localhost:5033)
 *   --auth-token <token>    Authentication token (required if using auth)
 * 
 * Example:
 *   node scripts/batch-csv-upload.js my-books.csv --batch-size 10 --delay 2000 --auth-token "your-jwt-token"
 */

const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');

// Parse command line arguments
function parseArgs() {
    const args = process.argv.slice(2);
    const config = {
        csvFile: null,
        batchSize: 10,
        delay: 1000,
        apiUrl: 'http://localhost:5033',
        authToken: null
    };

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        
        if (arg === '--batch-size' && args[i + 1]) {
            config.batchSize = parseInt(args[i + 1]);
            i++;
        } else if (arg === '--delay' && args[i + 1]) {
            config.delay = parseInt(args[i + 1]);
            i++;
        } else if (arg === '--api-url' && args[i + 1]) {
            config.apiUrl = args[i + 1];
            i++;
        } else if (arg === '--auth-token' && args[i + 1]) {
            config.authToken = args[i + 1];
            i++;
        } else if (!arg.startsWith('--') && !config.csvFile) {
            config.csvFile = arg;
        }
    }

    if (!config.csvFile) {
        console.error('Error: CSV file path is required');
        console.log('\nUsage: node batch-csv-upload.js <path-to-csv> [options]');
        process.exit(1);
    }

    if (!fs.existsSync(config.csvFile)) {
        console.error(`Error: File not found: ${config.csvFile}`);
        process.exit(1);
    }

    return config;
}

// Split CSV into batches
function splitCsvIntoBatches(csvContent, batchSize) {
    const lines = csvContent.split('\n');
    const header = lines[0];
    const dataLines = lines.slice(1).filter(line => line.trim() !== '');
    
    const batches = [];
    for (let i = 0; i < dataLines.length; i += batchSize) {
        const batchLines = dataLines.slice(i, i + batchSize);
        const batchCsv = [header, ...batchLines].join('\n');
        batches.push({
            number: Math.floor(i / batchSize) + 1,
            startRow: i + 1,
            endRow: Math.min(i + batchSize, dataLines.length),
            content: batchCsv,
            rowCount: batchLines.length
        });
    }
    
    return batches;
}

// Create a temporary CSV file for a batch
function createTempCsvFile(batchContent, batchNumber) {
    const tempDir = path.join(__dirname, 'temp-batches');
    if (!fs.existsSync(tempDir)) {
        fs.mkdirSync(tempDir, { recursive: true });
    }
    
    const tempFile = path.join(tempDir, `batch-${batchNumber}.csv`);
    fs.writeFileSync(tempFile, batchContent);
    return tempFile;
}

// Upload a batch using native http/https
function uploadBatch(config, batchFile, batchInfo) {
    return new Promise((resolve, reject) => {
        const boundary = '----WebKitFormBoundary' + Math.random().toString(36).substring(2);
        const fileContent = fs.readFileSync(batchFile);
        const fileName = path.basename(batchFile);
        
        // Build multipart form data
        let body = '';
        body += `--${boundary}\r\n`;
        body += `Content-Disposition: form-data; name="file"; filename="${fileName}"\r\n`;
        body += 'Content-Type: text/csv\r\n\r\n';
        
        const header = Buffer.from(body, 'utf8');
        const footer = Buffer.from(`\r\n--${boundary}--\r\n`, 'utf8');
        const bodyBuffer = Buffer.concat([header, fileContent, footer]);
        
        const url = new URL(`${config.apiUrl}/api/upload/csv`);
        const protocol = url.protocol === 'https:' ? https : http;
        
        const options = {
            hostname: url.hostname,
            port: url.port || (url.protocol === 'https:' ? 443 : 80),
            path: url.pathname,
            method: 'POST',
            headers: {
                'Content-Type': `multipart/form-data; boundary=${boundary}`,
                'Content-Length': bodyBuffer.length,
                ...(config.authToken && { 'Authorization': `Bearer ${config.authToken}` })
            }
        };
        
        const req = protocol.request(options, (res) => {
            let data = '';
            
            res.on('data', (chunk) => {
                data += chunk;
            });
            
            res.on('end', () => {
                try {
                    const response = JSON.parse(data);
                    if (res.statusCode >= 200 && res.statusCode < 300) {
                        resolve(response);
                    } else {
                        reject(new Error(`HTTP ${res.statusCode}: ${response.error || response.message || data}`));
                    }
                } catch (e) {
                    if (res.statusCode >= 200 && res.statusCode < 300) {
                        resolve({ message: 'Upload successful', rawResponse: data });
                    } else {
                        reject(new Error(`Failed to parse response: ${data}`));
                    }
                }
            });
        });
        
        req.on('error', (error) => {
            reject(error);
        });
        
        req.write(bodyBuffer);
        req.end();
    });
}

// Sleep utility
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// Main execution
async function main() {
    const config = parseArgs();
    
    console.log('='.repeat(60));
    console.log('Batch CSV Upload for ProjectLoopbreaker');
    console.log('='.repeat(60));
    console.log(`CSV File:       ${config.csvFile}`);
    console.log(`Batch Size:     ${config.batchSize} rows`);
    console.log(`Delay:          ${config.delay}ms`);
    console.log(`API URL:        ${config.apiUrl}`);
    console.log(`Auth:           ${config.authToken ? 'Yes (token provided)' : 'No'}`);
    console.log('='.repeat(60));
    console.log('');
    
    // Read and split CSV
    console.log('Reading CSV file...');
    const csvContent = fs.readFileSync(config.csvFile, 'utf8');
    const batches = splitCsvIntoBatches(csvContent, config.batchSize);
    
    console.log(`Total rows to process: ${batches.reduce((sum, b) => sum + b.rowCount, 0)}`);
    console.log(`Split into ${batches.length} batches`);
    console.log('');
    
    // Track results
    const results = {
        totalBatches: batches.length,
        successfulBatches: 0,
        failedBatches: 0,
        totalSuccessful: 0,
        totalErrors: 0,
        batchResults: []
    };
    
    // Process each batch
    for (const batch of batches) {
        console.log(`Processing batch ${batch.number}/${batches.length} (rows ${batch.startRow}-${batch.endRow})...`);
        
        const tempFile = createTempCsvFile(batch.content, batch.number);
        
        try {
            const response = await uploadBatch(config, tempFile, batch);
            
            results.successfulBatches++;
            results.totalSuccessful += (response.SuccessCount || response.successCount || 0);
            results.totalErrors += (response.ErrorCount || response.errorCount || 0);
            
            results.batchResults.push({
                batch: batch.number,
                success: true,
                successCount: response.SuccessCount || response.successCount || 0,
                errorCount: response.ErrorCount || response.errorCount || 0,
                message: response.Message || response.message
            });
            
            console.log(`  ✓ Success: ${response.SuccessCount || response.successCount || 0} items imported`);
            if ((response.ErrorCount || response.errorCount || 0) > 0) {
                console.log(`  ⚠ Warnings: ${response.ErrorCount || response.errorCount} errors in batch`);
                if (response.Errors && response.Errors.length > 0) {
                    response.Errors.slice(0, 3).forEach(err => console.log(`    - ${err}`));
                    if (response.Errors.length > 3) {
                        console.log(`    ... and ${response.Errors.length - 3} more errors`);
                    }
                }
            }
            
        } catch (error) {
            results.failedBatches++;
            results.batchResults.push({
                batch: batch.number,
                success: false,
                error: error.message
            });
            
            console.log(`  ✗ Failed: ${error.message}`);
        }
        
        // Clean up temp file
        fs.unlinkSync(tempFile);
        
        // Wait before next batch (except for the last one)
        if (batch.number < batches.length) {
            console.log(`  Waiting ${config.delay}ms before next batch...`);
            await sleep(config.delay);
            console.log('');
        }
    }
    
    // Clean up temp directory
    const tempDir = path.join(__dirname, 'temp-batches');
    if (fs.existsSync(tempDir)) {
        fs.rmdirSync(tempDir);
    }
    
    // Print summary
    console.log('');
    console.log('='.repeat(60));
    console.log('UPLOAD SUMMARY');
    console.log('='.repeat(60));
    console.log(`Total Batches:        ${results.totalBatches}`);
    console.log(`Successful Batches:   ${results.successfulBatches}`);
    console.log(`Failed Batches:       ${results.failedBatches}`);
    console.log(`Total Items Imported: ${results.totalSuccessful}`);
    console.log(`Total Errors:         ${results.totalErrors}`);
    console.log('='.repeat(60));
    
    if (results.failedBatches > 0) {
        console.log('\nFailed Batches:');
        results.batchResults
            .filter(r => !r.success)
            .forEach(r => {
                console.log(`  Batch ${r.batch}: ${r.error}`);
            });
    }
    
    console.log('\nDone!');
}

// Run the script
main().catch(error => {
    console.error('\nFatal Error:', error.message);
    process.exit(1);
});

File: scripts/batch-csv-upload.py
#!/usr/bin/env python3
"""
Batch CSV Upload Script for ProjectLoopbreaker

This script allows you to upload large CSV files in smaller batches
to avoid timeouts and make the process more manageable.

Usage:
    python scripts/batch-csv-upload.py <path-to-csv> [options]

Options:
    --batch-size <number>   Number of rows per batch (default: 10)
    --delay <seconds>       Delay between batches in seconds (default: 1)
    --api-url <url>         API base URL (default: http://localhost:5033)
    --auth-token <token>    Authentication token (required if using auth)

Example:
    python scripts/batch-csv-upload.py my-books.csv --batch-size 10 --delay 2 --auth-token "your-jwt-token"
"""

import argparse
import csv
import os
import sys
import time
import requests
from pathlib import Path
from typing import List, Dict, Any


def parse_args() -> argparse.Namespace:
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description='Batch CSV Upload for ProjectLoopbreaker',
        formatter_class=argparse.RawDescriptionHelpFormatter
    )
    
    parser.add_argument('csv_file', help='Path to the CSV file to upload')
    parser.add_argument('--batch-size', type=int, default=10,
                       help='Number of rows per batch (default: 10)')
    parser.add_argument('--delay', type=float, default=1.0,
                       help='Delay between batches in seconds (default: 1)')
    parser.add_argument('--api-url', default='http://localhost:5033',
                       help='API base URL (default: http://localhost:5033)')
    parser.add_argument('--auth-token', help='Authentication token')
    
    args = parser.parse_args()
    
    if not os.path.exists(args.csv_file):
        print(f"Error: File not found: {args.csv_file}")
        sys.exit(1)
    
    return args


def split_csv_into_batches(csv_file: str, batch_size: int) -> List[Dict[str, Any]]:
    """Split CSV file into batches."""
    batches = []
    
    with open(csv_file, 'r', encoding='utf-8') as f:
        reader = csv.reader(f)
        header = next(reader)
        rows = list(reader)
    
    # Filter out empty rows
    rows = [row for row in rows if any(cell.strip() for cell in row)]
    
    total_rows = len(rows)
    
    for i in range(0, total_rows, batch_size):
        batch_rows = rows[i:i + batch_size]
        batches.append({
            'number': (i // batch_size) + 1,
            'start_row': i + 1,
            'end_row': min(i + batch_size, total_rows),
            'header': header,
            'rows': batch_rows,
            'row_count': len(batch_rows)
        })
    
    return batches


def create_temp_csv_file(batch: Dict[str, Any], batch_number: int) -> str:
    """Create a temporary CSV file for a batch."""
    temp_dir = Path(__file__).parent / 'temp-batches'
    temp_dir.mkdir(exist_ok=True)
    
    temp_file = temp_dir / f'batch-{batch_number}.csv'
    
    with open(temp_file, 'w', newline='', encoding='utf-8') as f:
        writer = csv.writer(f)
        writer.writerow(batch['header'])
        writer.writerows(batch['rows'])
    
    return str(temp_file)


def upload_batch(api_url: str, batch_file: str, auth_token: str = None) -> Dict[str, Any]:
    """Upload a batch to the API."""
    url = f"{api_url}/api/upload/csv"
    
    headers = {}
    if auth_token:
        headers['Authorization'] = f'Bearer {auth_token}'
    
    with open(batch_file, 'rb') as f:
        files = {'file': (os.path.basename(batch_file), f, 'text/csv')}
        response = requests.post(url, files=files, headers=headers)
    
    response.raise_for_status()
    return response.json()


def main():
    """Main execution."""
    args = parse_args()
    
    print('=' * 60)
    print('Batch CSV Upload for ProjectLoopbreaker')
    print('=' * 60)
    print(f"CSV File:       {args.csv_file}")
    print(f"Batch Size:     {args.batch_size} rows")
    print(f"Delay:          {args.delay}s")
    print(f"API URL:        {args.api_url}")
    print(f"Auth:           {'Yes (token provided)' if args.auth_token else 'No'}")
    print('=' * 60)
    print()
    
    # Read and split CSV
    print('Reading CSV file...')
    batches = split_csv_into_batches(args.csv_file, args.batch_size)
    
    total_rows = sum(b['row_count'] for b in batches)
    print(f"Total rows to process: {total_rows}")
    print(f"Split into {len(batches)} batches")
    print()
    
    # Track results
    results = {
        'total_batches': len(batches),
        'successful_batches': 0,
        'failed_batches': 0,
        'total_successful': 0,
        'total_errors': 0,
        'batch_results': []
    }
    
    # Process each batch
    for batch in batches:
        print(f"Processing batch {batch['number']}/{len(batches)} "
              f"(rows {batch['start_row']}-{batch['end_row']})...")
        
        temp_file = create_temp_csv_file(batch, batch['number'])
        
        try:
            response = upload_batch(args.api_url, temp_file, args.auth_token)
            
            results['successful_batches'] += 1
            success_count = response.get('SuccessCount') or response.get('successCount') or 0
            error_count = response.get('ErrorCount') or response.get('errorCount') or 0
            
            results['total_successful'] += success_count
            results['total_errors'] += error_count
            
            results['batch_results'].append({
                'batch': batch['number'],
                'success': True,
                'success_count': success_count,
                'error_count': error_count,
                'message': response.get('Message') or response.get('message')
            })
            
            print(f"  ✓ Success: {success_count} items imported")
            if error_count > 0:
                print(f"  ⚠ Warnings: {error_count} errors in batch")
                errors = response.get('Errors', [])
                for err in errors[:3]:
                    print(f"    - {err}")
                if len(errors) > 3:
                    print(f"    ... and {len(errors) - 3} more errors")
            
        except requests.exceptions.RequestException as error:
            results['failed_batches'] += 1
            results['batch_results'].append({
                'batch': batch['number'],
                'success': False,
                'error': str(error)
            })
            
            print(f"  ✗ Failed: {error}")
        
        except Exception as error:
            results['failed_batches'] += 1
            results['batch_results'].append({
                'batch': batch['number'],
                'success': False,
                'error': str(error)
            })
            
            print(f"  ✗ Failed: {error}")
        
        finally:
            # Clean up temp file
            if os.path.exists(temp_file):
                os.remove(temp_file)
        
        # Wait before next batch (except for the last one)
        if batch['number'] < len(batches):
            print(f"  Waiting {args.delay}s before next batch...")
            time.sleep(args.delay)
            print()
    
    # Clean up temp directory
    temp_dir = Path(__file__).parent / 'temp-batches'
    if temp_dir.exists():
        try:
            temp_dir.rmdir()
        except OSError:
            pass  # Directory not empty or doesn't exist
    
    # Print summary
    print()
    print('=' * 60)
    print('UPLOAD SUMMARY')
    print('=' * 60)
    print(f"Total Batches:        {results['total_batches']}")
    print(f"Successful Batches:   {results['successful_batches']}")
    print(f"Failed Batches:       {results['failed_batches']}")
    print(f"Total Items Imported: {results['total_successful']}")
    print(f"Total Errors:         {results['total_errors']}")
    print('=' * 60)
    
    if results['failed_batches'] > 0:
        print('\nFailed Batches:')
        for result in results['batch_results']:
            if not result['success']:
                print(f"  Batch {result['batch']}: {result['error']}")
    
    print('\nDone!')


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('\n\nUpload interrupted by user.')
        sys.exit(1)
    except Exception as error:
        print(f'\nFatal Error: {error}')
        sys.exit(1)








